lib/krb5/principal.c

```diff
@@ -1260,17 +1260,43 @@ krb5_principal_is_anonymous(krb5_context context,
 {
     int anon_realm;
 
-    if ((p->name.name_type != KRB5_NT_WELLKNOWN &&
-         p->name.name_type != KRB5_NT_UNKNOWN) ||
-        p->name.name_string.len != 2 ||
+    /*
+     * Heimdal versions 7.5 and below left the name-type at KRB5_NT_PRINCIPAL
+     * even with anonymous pkinit responses.  To retain interoperability with
+     * legacy KDCs, the name-type is not checked by the client after requesting
+     * a fully anonymous ticket.
+     */
+    if ((flags & KRB5_ANON_MATCH_NAME_TYPE) &&
+        p->name.name_type != KRB5_NT_WELLKNOWN &&
+        p->name.name_type != KRB5_NT_UNKNOWN)
+        return FALSE;
+
+    if (p->name.name_string.len != 2 ||
         strcmp(p->name.name_string.val[0], KRB5_WELLKNOWN_NAME) != 0 ||
         strcmp(p->name.name_string.val[1], KRB5_ANON_NAME) != 0)
         return FALSE;
 
-    anon_realm = strcmp(p->realm, KRB5_ANON_REALM) == 0;
+    /*
+     * While unauthenticated clients SHOULD get "WELLKNOWN:ANONYMOUS" as their
+     * realm, Heimdal KDCs prior to 7.0 returned the requested realm.  While
+     * such tickets might lead *servers* to unwittingly grant access to fully
+     * anonymous clients, trusting that the client was authenticated to the
+     * realm in question, doing it right is the KDCs job, the client should not
+     * refuse such a ticket.
+     *
+     * If we ever do decide to enforice WELLKNOWN:ANONYMOUS for unathenticted
+     * clients, it is essential that calls that pass KRB5_ANON_MATCH_ANY still
+     * ignore the realm, as in that case either case matches one of the two
+     * possible conditions.
+     */
+    if (flags & KRB5_ANON_MATCH_UNAUTHENTICATED)
+        return TRUE;
```

Comment:
Hmm, I don't know about this. Other (not merged) code such as the `GSS_C_ANON_FLAG` support relies on this function. I would prefer to put the backwards compatibility logic into the caller and have `krb5_principal_is_anonymous()` be RFC compliant.

Reply:
Most of the other comments seem reasonable, and easy to fix.  Here I disagree.  If this function is going to enforce the name type, then what do you suggest a caller that wants to interoperate with &le; 7.5 Heimdal KDCs should do?  The function becomes unusable and we need a new one? Or what?

Reply:
I will need to think it through more carefully. My gut feel is that –

* the function should by default implement the RFC compliant behavior – the UNAUTHENTICATED test matches unauthenticated only, and the AUTHENTICATED test matches authenticated only
* if compatibility behavior is desired, it should preferably be set by the caller with additional flags, or failing that, krb5.conf should be checked within the function

FWIW, the function was originally supposed to be an internal function but it was exposed because it was difficult to expose the flags without adding them to krb5.h (krb5_locl.h can't be easily imported outside lib/krb5). And if the flags are to be public, then we thought so should the function.

Reply:
I still think the client is not the right place to enforce lax policy by the KDC.  I'm leaving this as-is for now.